### PR TITLE
feat(sequencer): implement transaction fee query

### LINKED ISSUE #1071: sequencer: `estimate_fee` api endpoint
implement a sequencer API endpoint which returns the fee for a given transaction (based on the current head state)

### PR DESCRIPTION
**Pull request by noot**
## Summary
implement transaction fee ABCI query method. calculates the fee for the transaction based on the current state.

## Background
useful for UX and fee estimation.

## Changes
- implement transaction fee ABCI query method which calculates the fee for the transaction based on the current state
- update sequencer-client respectively

## Testing
unit tests

## Related Issues

closes #1071

### CODE REVIEW

File: crates/astria-sequencer/src/transaction/query.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    };
+
+    let fees = match get_fees_for_transaction(&tx, &snapshot).await {
+        Ok(fees) => fees,
+        Err(err) => {
+            return response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
+                log: format!("failed calculating fees for provided transaction: {err:?}"),
```
Comment (Fraser999) [inline crates/astria-sequencer/src/transaction/query.rs:52]:
```suggestion
                log: format!("failed calculating fees for provided transaction: {err:#}"),
```

File: crates/astria-sequencer/src/transaction/query.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        height,
+        fees,
+    };
+
+    let payload = resp.into_raw().encode_to_vec().into();
+
+    let height = tendermint::block::Height::try_from(height).expect("height must fit into an i64");
+    response::Query {
+        code: 0.into(),
+        key: request.path.clone().into_bytes().into(),
```
Comment (Fraser999) [inline crates/astria-sequencer/src/transaction/query.rs:70]:
```suggestion
        key: request.path.into_bytes().into(),
```

File: crates/astria-sequencer/src/transaction/query.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+}
+
+fn preprocess_request(request: &request::Query) -> Result<UnsignedTransaction, response::Query> {
+    let tx = match RawUnsignedTransaction::decode(&*request.data) {
+        Ok(tx) => tx,
+        Err(err) => {
+            return Err(response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
```
Comment (Fraser999) [inline crates/astria-sequencer/src/transaction/query.rs:83]: This doesn't seem like an internal server error - it points to the request being malformed.  I'm not sure if any of the existing `AbciErrorCode` variants really fit.  We need something like a `BAD_REQUEST` (equivalent to HTTP 400 code) variant I think.

File: crates/astria-sequencer/src/transaction/query.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            });
+        }
+    };
+
+    let tx = match UnsignedTransaction::try_from_raw(tx) {
+        Ok(tx) => tx,
+        Err(err) => {
+            return Err(response::Query {
+                code: AbciErrorCode::INTERNAL_ERROR.into(),
+                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
```
Comment (Fraser999) [inline crates/astria-sequencer/src/transaction/query.rs:95]: Same for this case - more like a `BAD_REQUEST` error?

**Review by Fraser999 — APPROVED** [review]

File: crates/astria-core/src/protocol/abci.rs [hunk 5]
```diff
@@ -37,6 +38,7 @@ impl AbciErrorCode {
             8 => "the requested value was not found".into(),
             9 => "the transaction expired in the app's mempool".into(),
             10 => "the transaction failed to execute in prepare_proposal()".into(),
+            11 => "the request was malformed".into(),
```
Comment (SuperFluffy) [inline crates/astria-core/src/protocol/abci.rs:41]: Maybe "the request body" or "request payload"?

File: crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs [hunk 6]
```diff
@@ -470,6 +473,83 @@ impl TransactionParams {
...
+    fn asset(inner: asset::ParseDenomError) -> Self {
+        Self(TransactionFeeResponseErrorKind::Asset(inner))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum TransactionFeeResponseErrorKind {
+    #[error("`fee` field is unset")]
+    UnsetFee,
+    #[error("failed to parse asset denom")]
```
Comment (SuperFluffy) [inline crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs:549]: Maybe add the field, i.e. `failed to parse ok asset listed in the ``.assets`` field`

File: crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs [hunk 7]
```diff
@@ -470,6 +473,83 @@ impl TransactionParams {
...
+    Asset(#[from] asset::ParseDenomError),
```
Comment (SuperFluffy) [inline crates/astria-core/src/protocol/transaction/v1alpha1/mod.rs:550]: Better just `#[source]` because I think we don't use the from conversion ever

File: crates/astria-sequencer/src/transaction/checks.rs [hunk 8]
```diff
@@ -57,17 +57,15 @@ pub(crate) async fn check_balance_mempool<S: StateReadExt + 'static>(
...
 
-// Checks that the account has enough balance to cover the total fees and transferred values
-// for all actions in the transaction.
-pub(crate) async fn check_balance_for_total_fees<S: StateReadExt + 'static>(
+pub(crate) async fn get_fees_for_transaction<S: StateReadExt + 'static>(
     tx: &UnsignedTransaction,
-    from: Address,
     state: &S,
-) -> anyhow::Result<()> {
+) -> anyhow::Result<HashMap<asset::IbcPrefixed, u128>> {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/checks.rs:68]: Fine for doing this in a follow-up: cant we request all the data bits in this method concurrently instead of one after the other?

File: crates/astria-sequencer/src/transaction/checks.rs [hunk 9]
```diff
@@ -148,7 +132,64 @@ pub(crate) async fn check_balance_for_total_fees<S: StateReadExt + 'static>(
...
+            | Action::Ibc(_)
+            | Action::IbcRelayerChange(_)
+            | Action::FeeAssetChange(_)
+            | Action::FeeChange(_) => {
+                continue;
+            }
+        }
+    }
+
+    for (asset, total_fee) in cost_by_asset {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/checks.rs:192]: Could fire these up concurrently too

File: proto/protocolapis/astria/protocol/transactions/v1alpha1/types.proto [hunk 10]
```diff
@@ -246,3 +246,14 @@ message FeeChangeAction {
     astria.primitive.v1.Uint128 ics20_withdrawal_base_fee = 40;
   }
 }
+
+// Response to a transaction fee ABCI query.
+message TransactionFeeResponse {
+  uint64 height = 2;
```
Comment (SuperFluffy) [inline proto/protocolapis/astria/protocol/transactions/v1alpha1/types.proto:252]: Index from 1?

File: crates/astria-sequencer/src/transaction/checks.rs [hunk 11]
```diff
@@ -164,16 +205,10 @@ pub(crate) async fn check_balance_for_total_fees<S: StateReadExt + 'static>(
 }
 
 fn transfer_update_fees(
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/checks.rs:207]: Why is this body updated? Probably missing something, but from a cursory glance this looks orthogonal to the new functionality?

File: crates/astria-sequencer/src/transaction/checks.rs [hunk 12]
```diff
@@ -197,16 +232,10 @@ async fn sequence_update_fees<S: StateReadExt>(
 }
 
 fn ics20_withdrawal_updates_fees(
-    asset: &asset::Denom,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/checks.rs:200]: Same here - is this functionality no longer necessary or wrong?

File: crates/astria-sequencer/src/transaction/checks.rs [hunk 13]
```diff
@@ -233,37 +262,21 @@ fn bridge_lock_update_fees(
...
         .entry(act.asset.to_ibc_prefixed())
         .and_modify(|amt| *amt = amt.saturating_add(expected_deposit_fee))
         .or_insert(expected_deposit_fee);
 }
 
-async fn bridge_unlock_update_fees<S: StateReadExt>(
-    state: &S,
-    bridge_address: Address,
-    amount: u128,
+fn bridge_unlock_update_fees(
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/checks.rs:271]: Same here - why all that change?

**Review by SuperFluffy — COMMENTED** [review]
I left comments looking to understand why various code was updated (updating fees etc), even the PR just aims to read (not write)?

File: crates/astria-sequencer/src/transaction/checks.rs (hunk 11 above)
Comment (noot) [reply crates/astria-sequencer/src/transaction/checks.rs:207]: the function checking "fees" before was actually checking the total value used by the transaction, so it was summing fees and value transfers. however, the fees endpoint wants to return only fees used by the tx, so i had to separate these

File: crates/astria-sequencer/src/transaction/checks.rs (hunk 8 above)
Comment (noot) [reply crates/astria-sequencer/src/transaction/checks.rs:68]: yes - can make an issue

File: proto/protocolapis/astria/protocol/transactions/v1alpha1/types.proto (hunk 10 above)
Comment (noot) [reply proto/protocolapis/astria/protocol/transactions/v1alpha1/types.proto:252]: we should probably change all of the response types if we want to index from 1, maybe we can do that when we move them to their own package?

File: crates/astria-sequencer/src/transaction/checks.rs (hunk 12 above)
Comment (noot) [reply crates/astria-sequencer/src/transaction/checks.rs:200]: same as above

File: crates/astria-sequencer/src/transaction/checks.rs (hunk 13 above)
Comment (noot) [reply crates/astria-sequencer/src/transaction/checks.rs:271]: same as above

File: crates/astria-sequencer/src/transaction/checks.rs [hunk 14]
```diff
@@ -57,17 +57,15 @@ pub(crate) async fn check_balance_mempool<S: StateReadExt + 'static>(
     state: &S,
 ) -> anyhow::Result<()> {
     let signer_address = crate::address::base_prefixed(tx.verification_key().address_bytes());
-    check_balance_for_total_fees(tx.unsigned_transaction(), signer_address, state).await?;
+    check_balance_for_total_fees_and_transfers(tx.unsigned_transaction(), signer_address, state)
+        .await?;
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/checks.rs:61]: context

File: crates/astria-sequencer/src/transaction/checks.rs [hunk 15]
```diff
@@ -148,7 +132,64 @@ pub(crate) async fn check_balance_for_total_fees<S: StateReadExt + 'static>(
...
+}
+
+// Checks that the account has enough balance to cover the total fees and transferred values
+// for all actions in the transaction.
+pub(crate) async fn check_balance_for_total_fees_and_transfers<S: StateReadExt + 'static>(
+    tx: &UnsignedTransaction,
+    from: Address,
+    state: &S,
+) -> anyhow::Result<()> {
+    let mut cost_by_asset = get_fees_for_transaction(tx, state).await?;
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/checks.rs:145]: context

File: crates/astria-sequencer/src/transaction/mod.rs [hunk 16]
```diff
@@ -198,7 +199,7 @@ impl ActionHandler for UnsignedTransaction {
         ensure!(curr_nonce == self.nonce(), InvalidNonce(self.nonce()));
 
         // Should have enough balance to cover all actions.
-        check_balance_for_total_fees(self, from, state).await?;
+        check_balance_for_total_fees_and_transfers(self, from, state).await?;
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/transaction/mod.rs:202]: context

**Review by SuperFluffy — APPROVED** [review]
approve for API.

few comments on missing error context
